src/mkdocs_gallery/gen_single.py

```diff
@@ -739,6 +739,66 @@ def _reset_cwd_syspath(cwd, path_to_remove):
     os.chdir(cwd)
 
 
+def _parse_code(bcontent, src_file, *, compiler_flags):
+    code_ast = compile(bcontent, src_file, "exec", compiler_flags | ast.PyCF_ONLY_AST, dont_inherit=1)
+    if _needs_async_handling(bcontent, compiler_flags=compiler_flags):
+        code_ast = _apply_async_handling(code_ast, compiler_flags=compiler_flags)
+    return code_ast
+
+
+def _needs_async_handling(bcontent, *, compiler_flags) -> bool:
+    try:
+        compile(bcontent, "<_needs_async_handling()>", "exec", compiler_flags, dont_inherit=1)
+    except SyntaxError as error:
+        # FIXME
+        #  bool(re.match(r"'(await|async for|async with)' outside( async)? function", str(error)))
+        #  asynchronous comprehension outside of an asynchronous function
+        return "async" in str(error)
```

Comment:
As far as I'm aware there are four error messages that we need to look out for:
- await outside function
- async for outside async function
- async with outside async function
- asynchronous comprehension outside of an asynchronous function

Kind of a bummer that they don't follow the same pattern, but here we are. We can still fit this into a reasonable regex.

A simpler, but less robust solution (and yes I'm aware that using the error message as trigger is not robust in the first place), would be to just check for `"async"` as implemented right now. LMK what you think.

Reply:
From what I understand from https://github.com/ipython/ipython/pull/10390/files#diff-6abb45111ba17c2eb57fa3302b2e73aec68a5adfa6805dcf90d3fe1e9e72bd69R61

Their solution is even simpler: they try to apply the async patch, and if it works then it means that it was needed :)

Could the same work for us ?

Reply:
Sure, we can do that as well. Let me fix it.

Reply:
Implemented in 5e0e63693b5c951687d7f8519ec253a84d200ca8.